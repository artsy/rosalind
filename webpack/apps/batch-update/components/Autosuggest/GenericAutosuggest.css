.GenericAutosuggest input {
  width: 100%;
  margin: 1em 0 0 0;
  line-height: 2em;
  font-size: 1em;
  border: none;
  border-bottom: solid 1px #ddd;
  background: none;
}

.GenericAutosuggest input:focus {
  outline: none;
}

.GenericAutosuggest ul.react-autosuggest__suggestions-list {
  list-style-type: none;
}

.GenericAutosuggest li.react-autosuggest__suggestion {
  display: flex;
  align-items: center;
  min-height: 2em;
  padding: 0 0.25em;
  background: white;
}

.GenericAutosuggest li.react-autosuggest__suggestion--focused {
  background: purple;
  color: white;
}
