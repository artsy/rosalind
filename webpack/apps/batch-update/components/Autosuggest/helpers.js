import React from 'react'

export const getSuggestionValue = (suggestion) => suggestion.name
export const renderSuggestion = (suggestion) => <div>{suggestion.name}</div>
