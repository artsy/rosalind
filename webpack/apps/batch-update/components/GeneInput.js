import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { colors } from './Layout'

class GeneInput extends React.Component {
  constructor (props) {
    super(props)
    this.handleChange = this.handleChange.bind(this)
  }

  handleChange (e) {
    const { onChangeValue, name } = this.props
    const value = e.target.value
    onChangeValue({ name, value })
  }

  render () {
    const { name, value } = this.props
    return (
      <Square value={value}>
        <Name value={value}>
          { value === 0 ? <del>{name}</del> : name }
          <Bar value={value} />
        </Name>
        <Value
          type='number'
          min='0'
          max='100'
          step='10'
          placeholder='Multiple'
          value={value === null ? '' : value}
          onChange={this.handleChange}
        />
      </Square>
    )
  }
}

GeneInput.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.number,
  onChangeValue: PropTypes.func
}

const Square = styled.div`
  ${props => {
    if (parseInt(props.value) === 0) {
      return `
        border: solid 1px ${colors.red};
      `
    } else if (props.value) {
      return `
        border: solid 1px ${colors.purple};
      `
    } else {
      return `
        border: solid 1px ${colors.grayLighter};
      `
    }
  }}
  width: 140px;
  height: 140px;
  margin-right: 20px;
  margin-bottom: 20px;
  display: flex;
  flex-direction: column;
`
Square.displayName = 'Square'

const Name = styled.div`
  ${props => {
    if (parseInt(props.value) === 0) {
      return `
        background: ${colors.redLightest};
        border-bottom: solid 1px ${colors.red};
      `
    } else if (props.value) {
      return `
        background: ${colors.purpleLightest};
        border-bottom: solid 1px ${colors.purple};
      `
    } else {
      return `
        background: ${colors.grayLightest};
        border-bottom: solid 1px ${colors.grayLighter};
      `
    }
  }}
  flex-basis: 100px;
  padding: 0.5em 1em;
  overflow: hidden;
  position: relative;
`
Name.displayName = 'Name'

const Value = styled.input`
  flex-basis: 40px;
  border-color: transparent;
  padding: 1em;
  &:focus {
    outline: solid 1px ${colors.gray};
  }
`
Value.displayName = 'Value'

const Bar = styled.div`
  position: absolute;
  left: 0;
  bottom: 0;
  width: 100%;
  background: hsla(261, 100%, 85%, 0.15);
  transition: height 0.15s;
  ${props => {
    return `
      height: ${props.value}%;
    `
  }}
`
Bar.displayName = 'Bar'

export default GeneInput
