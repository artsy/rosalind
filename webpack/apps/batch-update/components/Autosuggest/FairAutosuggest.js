import React from 'react'
import PropTypes from 'prop-types'
import GenericAutosuggest from './GenericAutosuggest'
import { matchFairs } from 'lib/rosalind-api'
import { getSuggestionValue, renderSuggestion } from './helpers'

function FairAutosuggest (props) {
  return (
    <GenericAutosuggest
      id='fair-autosuggest'
      placeholder={props.placeholder}
      fetchSuggestions={matchFairs}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      selectSuggestion={fair => { props.updateState('fair', fair) }}
    />
  )
}

FairAutosuggest.propTypes = {
  updateState: PropTypes.func
}

FairAutosuggest.defaultProps = {
  placeholder: 'Select a fair'
}

export { FairAutosuggest }
