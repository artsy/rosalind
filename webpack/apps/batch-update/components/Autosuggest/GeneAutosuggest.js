import React from 'react'
import GenericAutosuggest from './GenericAutosuggest'
import { matchGenes } from 'lib/rosalind-api'
import { getSuggestionValue, renderSuggestion } from './helpers'

function GeneAutosuggest (props) {
  return (
    <GenericAutosuggest
      id='gene-autosuggest'
      placeholder={props.placeholder}
      fetchSuggestions={matchGenes}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      selectSuggestion={gene => { props.onSelectGene(gene) }}
      />
  )
}

GeneAutosuggest.propTypes = {
  onSelectGene: React.PropTypes.func.isRequired
}

GeneAutosuggest.defaultProps = {
  placeholder: 'Select a gene'
}

export { GeneAutosuggest }
