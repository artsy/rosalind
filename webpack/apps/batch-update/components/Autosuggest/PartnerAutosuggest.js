import React from 'react'
import GenericAutosuggest from './GenericAutosuggest'
import { matchPartners } from 'lib/rosalind-api'
import { getSuggestionValue, renderSuggestion } from './helpers'

function PartnerAutosuggest (props) {
  return (
    <GenericAutosuggest
      id='partner-autosuggest'
      placeholder={props.placeholder}
      fetchSuggestions={matchPartners}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      selectSuggestion={partner => { props.onSelectPartner(partner) }}
      />
  )
}

PartnerAutosuggest.propTypes = {
  onSelectPartner: React.PropTypes.func.isRequired
}

PartnerAutosuggest.defaultProps = {
  placeholder: 'Select a partner'
}

export { PartnerAutosuggest }
