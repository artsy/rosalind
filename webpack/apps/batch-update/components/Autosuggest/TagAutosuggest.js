import React from 'react'
import GenericAutosuggest from './GenericAutosuggest'
import { matchTags } from 'lib/rosalind-api'
import { getSuggestionValue, renderSuggestion } from './helpers'

function TagAutosuggest (props) {
  return (
    <GenericAutosuggest
      id='tag-autosuggest'
      placeholder={props.placeholder}
      fetchSuggestions={matchTags}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      selectSuggestion={tag => { props.onSelectTag(tag) }}
      />
  )
}

TagAutosuggest.propTypes = {
  onSelectTag: React.PropTypes.func.isRequired
}

TagAutosuggest.defaultProps = {
  placeholder: 'Select a tag'
}

export { TagAutosuggest }
