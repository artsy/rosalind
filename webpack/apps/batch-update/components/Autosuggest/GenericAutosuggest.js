import React from 'react'
import Autosuggest from 'react-autosuggest'
import './GenericAutosuggest.css'

export default class GenericAutosuggest extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      value: '',
      suggestions: []
    }
    this.onSuggestionsFetchRequested = this.onSuggestionsFetchRequested.bind(this)
    this.onSuggestionsClearRequested = this.onSuggestionsClearRequested.bind(this)
    this.onChange = this.onChange.bind(this)
    this.onSuggestionSelected = this.onSuggestionSelected.bind(this)
  }

  onSuggestionsFetchRequested ({value}) {
    this.props.fetchSuggestions(value).then(suggestions => {
      this.setState({ suggestions })
    })
  }

  onSuggestionsClearRequested () {
    this.setState({
      suggestions: []
    })
  }

  onChange (event, { newValue }) {
    this.setState({
      value: newValue
    })
  }

  onSuggestionSelected (event, { suggestion }) {
    this.props.selectSuggestion(suggestion)
    this.setState({value: ''})
  }

  render () {
    const inputElementProps = {
      placeholder: this.props.placeholder,
      value: this.state.value,
      onChange: this.onChange
    }

    return (
      <div className='GenericAutosuggest'>
        <Autosuggest
          id={this.props.id}
          suggestions={this.state.suggestions}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          getSuggestionValue={this.props.getSuggestionValue}
          renderSuggestion={this.props.renderSuggestion}
          inputProps={inputElementProps}
          onSuggestionSelected={this.onSuggestionSelected}
          />
      </div>
    )
  }
}

GenericAutosuggest.propTypes = {
  id: React.PropTypes.string,
  placeholder: React.PropTypes.string,
  fetchSuggestions: React.PropTypes.func.isRequired,   // searchTerm => listOfMatchingSuggestionObjects
  getSuggestionValue: React.PropTypes.func.isRequired, // suggestionObject => displayName
  renderSuggestion: React.PropTypes.func.isRequired,   // suggestionObject => stringOrMarkupForSuggestionList
  selectSuggestion: React.PropTypes.func.isRequired     // suggestionObject => { handlerFunction(suggestionObject) }
}
