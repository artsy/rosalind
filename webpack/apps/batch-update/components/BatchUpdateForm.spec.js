import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { mount } from 'enzyme'
import BatchUpdateForm from './BatchUpdateForm'
import GeneInput from './GeneInput'
import { GeneAutosuggest } from './Autosuggest'
import ConfirmationModal from './ConfirmationModal'
import * as rosalindApi from 'lib/rosalind-api'

let props, dismissHandler

beforeEach(() => {
  dismissHandler = jest.fn()
  props = {
    onCancel: dismissHandler,
    selectedArtworkIds: ['one', 'two', 'three'],
    getCommonGenes: jest.fn().mockReturnValueOnce(['Art', 'Painting']),
    onAddNotice: jest.fn()
  }
})

it('renders correctly', () => {
  const rendered = renderer.create(<BatchUpdateForm {...props} />)
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the common genes for each new artwork selection', () => {
  const wrapper = mount(<BatchUpdateForm {...props} />)
  expect(wrapper.state().geneValues).toEqual({})
  wrapper.setProps({
    selectedArtworkIds: ['one', 'two', 'three', 'four']
  }) // triggers componentWillReceiveProps()
  expect(wrapper.state().geneValues).toMatchObject({
    'Art': null,
    'Painting': null
  })
  expect(wrapper.find(GeneInput).length).toEqual(2)
})

describe('when the "Cancel" link is clicked', () => {
  it('calls the correct handler', () => {
    const wrapper = mount(<BatchUpdateForm {...props} />)
    const mockClickEvent = { preventDefault: jest.fn() }
    wrapper.find('a.cancel').simulate('click', mockClickEvent)
    expect(dismissHandler.mock.calls.length).toEqual(1)
  })

  it('resets the form state', () => {
    const wrapper = mount(<BatchUpdateForm {...props} />)
    wrapper.setState({
      geneValues: {
        'Kawaii': 70
      }
    })
    const mockClickEvent = { preventDefault: jest.fn() }
    wrapper.find('a.cancel').simulate('click', mockClickEvent)
    expect(wrapper.state('geneValues')).toEqual({
      'Art': null,
      'Painting': null
    })
  })
})

it('adds genes', () => {
  const wrapper = mount(<BatchUpdateForm {...props} />)
  const component = wrapper.instance()
  component.onAddGene({ id: 'kawaii', name: 'Kawaii' })
  expect(component.state).toMatchObject({
    geneValues: {
      'Kawaii': null
    }
  })
})

describe('with no currently added genes', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(<BatchUpdateForm {...props} />)
    wrapper.setState({
      geneValues: {}
    })
  })

  it('disables the "Queue" button', () => {
    expect(wrapper.find('button.queue').props().disabled).toEqual(true)
  })
})

describe('with only null genes', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(<BatchUpdateForm {...props} />)
    wrapper.setState({
      geneValues: {
        'Kawaii': null
      }
    })
  })

  it('disables the "Queue" button', () => {
    expect(wrapper.find('button.queue').props().disabled).toEqual(true)
  })
})

describe('with currently added genes', () => {
  let wrapper, component

  beforeEach(() => {
    wrapper = mount(<BatchUpdateForm {...props} />)
    component = wrapper.instance()
    component.setState({
      geneValues: {
        'Kawaii': 0,
        'Animals': 100
      }
    })
  })

  it('enables the "Queue" button', () => {
    expect(wrapper.find('button.queue').props().disabled).toEqual(false)
  })

  it('renders the current genes', () => {
    expect(wrapper.find(GeneInput).length).toEqual(2)
    expect(wrapper.text()).toMatch('Kawaii')
    expect(wrapper.text()).toMatch('Animals')
  })

  it('updates gene values from integers', () => {
    component.onChangeGeneValue({ name: 'Kawaii', value: 70 })
    expect(component.state['geneValues']).toMatchObject({
      'Kawaii': 70
    })
  })

  it('updates gene values from number-ish strings', () => {
    component.onChangeGeneValue({ name: 'Kawaii', value: '70' })
    expect(component.state['geneValues']).toMatchObject({
      'Kawaii': 70
    })
  })

  it('nulls gene values from empty strings', () => {
    component.onChangeGeneValue({ name: 'Kawaii', value: '' })
    expect(component.state['geneValues']).toMatchObject({
      'Kawaii': null
    })
  })

  describe('when the "Queue" button is clicked', () => {
    it('opens a confirmation modal', () => {
      const mockClickEvent = { preventDefault: jest.fn() }
      wrapper.find('button.queue').simulate('click', mockClickEvent)
      expect(wrapper.find(ConfirmationModal).hasClass('modal-open')).toBe(true)
    })

    describe('when the confirmation modal is accepted', () => {
      beforeAll(() => {
        // We assume the existence of a <meta> tag containing the
        // Rails-generated CSRF token. This is how we mock access to
        // that tag.
        // https://github.com/facebook/jest/issues/2297
        Object.defineProperty(document, 'querySelector', {
          value: () => { return { content: 'very secret csrf token' } }
        })
      })

      beforeEach(() => {
        // mock the function that submits the request to the backend and returns
        // window.fetch's promise, to be handled by the client
        rosalindApi.submitBatchUpdate = jest.fn((artworks, genes, token) => {
          return new Promise((resolve, reject) => {
            resolve({ ok: true })
          })
        })

        // suppress log output in test run output
        console.log = jest.fn()
        console.error = jest.fn()

        const mockClickEvent = { preventDefault: jest.fn() }
        wrapper.find('button.queue').simulate('click', mockClickEvent)
        wrapper.find(ConfirmationModal).find('a.accept').simulate('click', mockClickEvent)
      })

      it('submits the form', () => {
        expect(rosalindApi.submitBatchUpdate.mock.calls.length).toEqual(1)
      })

      it('sends the genes and artworks', () => {
        expect(rosalindApi.submitBatchUpdate.mock.calls[0][0]).toEqual(props.selectedArtworkIds)
        expect(rosalindApi.submitBatchUpdate.mock.calls[0][1]).toEqual({'Animals': 100, 'Kawaii': 0})
      })

      it('includes the forgery token from the html document', () => {
        expect(rosalindApi.submitBatchUpdate.mock.calls[0][2]).toEqual('very secret csrf token')
      })
    })
  })
})

it('includes a gene autocomplete widget', () => {
  const wrapper = mount(<BatchUpdateForm {...props} />)
  expect(wrapper.find(GeneAutosuggest).length).toEqual(1)
})
