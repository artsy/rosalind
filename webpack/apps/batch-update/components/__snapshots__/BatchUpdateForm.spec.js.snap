exports[`test renders correctly 1`] = `
<div
  className="igHpwR">
  <div
    className="gnEZXA">
    <a
      className="cancel gHceZM"
      href="#"
      onClick={[Function]}>
      Cancel
    </a>
    <div>
      3
       works selected
    </div>
    <button
      className="queue feplCa"
      disabled={true}
      onClick={[Function]}>
      Queue changes
    </button>
  </div>
  <div
    className="hfTCBQ" />
  <div
    className="kXtoxa">
    There arenâ€™t any genes that describe all of your selected works
  </div>
  <div
    className="GenericAutosuggest">
    <div
      className="react-autosuggest__container">
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        aria-haspopup={false}
        aria-owns="react-autowhatever-gene-autosuggest"
        autoComplete="off"
        className="react-autosuggest__input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Add a gene"
        role="combobox"
        type="text"
        value="" />
      <div
        className="react-autosuggest__suggestions-container"
        id="react-autowhatever-gene-autosuggest" />
    </div>
  </div>
  <div
    className="mluQ" />
  <div
    className="gungoz">
    <h1>
      Are you sure you want to queue these changes?
    </h1>
    <section>
      <p>
        You will be changing the genome of 
        3
         works
      </p>
    </section>
    <div
      className="kriMBh">
      <a
        className="dismiss jRkNFJ"
        href="#"
        onClick={[Function]}>
        Go back
      </a>
      <a
        className="accept jhxkip"
        href="#"
        onClick={[Function]}>
        Continue
      </a>
    </div>
  </div>
</div>
`;
